### PR TITLE
feat: Added `Bits<T>` in num module

### PR DESCRIPTION
**Pull request by akhercha**
Resolves: #4245 

## Added

- `Representation<T>` trait & its implementation for scalar types.

## Comments

I'm not 100% sure for the naming ; happy to change if you have something better!

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)

### CODE REVIEW

**Review by orizi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @akhercha)

___
*[`corelib/src/lib.cairo` line 338 at r1](https://reviewable.io/reviews/starkware-libs/cairo/4251#-NgYUgFh1c5qcgVEc49W:-NgYUgFh1c5qcgVEc49X:b-67ra1s) ([raw file](https://github.com/starkware-libs/cairo/blob/4e9038ca07b7ce9f67d6a4a7c0e50d5f0084dd3e/corelib/src/lib.cairo#L338)):*
> ```Cairo
> 
> // Representation.
> mod repr;
> ```

Find a better name for all.
This sounds like python repr function.



_Code quote:_
```Cairo
// Representation.
mod repr;
```
___
*[`corelib/src/repr.cairo` line 2 at r1](https://reviewable.io/reviews/starkware-libs/cairo/4251#-NgYUxg_D_Nkymf4jjKY:-NgYUxga39vJf1dOCKCg:buy9fk) ([raw file](https://github.com/starkware-libs/cairo/blob/4e9038ca07b7ce9f67d6a4a7c0e50d5f0084dd3e/corelib/src/repr.cairo#L2)):*
> ```Cairo
> // === Representation ===
> ```

remove all inlines - specifically unrequired, and generally avoid.

**Comment by enitrat** [comment]
I don't think the name `representation` carries much semantics on its own. Could we try to find a more meaningful name, like `BitLength`, `BitSize` or `BitCapacity` - or something else?

**Review by Eikix — COMMENTED** [review]
I like what this pr enables us to do, most notably for kakarot in bitshifts,
i'm not aware of how this is achieved in other programming languages, but 'bits_repr' isn't so expressive

**Comment by akhercha** [comment]
> I don't think the name `representation` carries much semantics on its own. Could we try to find a more meaningful name, like `BitLength`, `BitSize` or `BitCapacity` - or something else (BitRepresentation shortened to BitRepr?)?

Yea this was mainly for """extandability""", I had in mind that more "representations" (kinda means nothing yea) could have be added to it later if needed.

As you suggested it would be something like:

```rust
trait Capacity<T> {
    fn bits() -> T;                                    // U8Capacity::bits()
    fn bits_capacityr(self: @T) -> T;    // 1_u8::bits_capacity()
}
```

or

```rust
trait Length<T> {
    fn bits() -> T;                                   // U8Length::bits()
    fn bits_length(self: @T) -> T;       // 1_u8::bits_length()
}
```

Any preference @enitrat ?

**Comment by Eikix** [comment]
> > I don't think the name `representation` carries much semantics on its own. Could we try to find a more meaningful name, like `BitLength`, `BitSize` or `BitCapacity` - or something else (BitRepresentation shortened to BitRepr?)?
> 
> Yea this was mainly for """extandability""", I had in mind that more "representations" (kinda means nothing yea) could have be added to it later if needed.
> 
> As you suggested it would be something like:
> 
> ```rust
> trait Capacity<T> {
>     fn bits() -> T;                                    // U8Capacity::bits()
>     fn bits_capacity(self: @T) -> T;    // 1_u8::bits_capacity()
> }
> ```
> 
> or
> 
> ```rust
> trait Length<T> {
>     fn bits() -> T;                                   // U8Length::bits()
>     fn bits_length(self: @T) -> T;       // 1_u8::bits_length()
> }
> ```
> 
> Any preference @enitrat ? Or I can keep `Representation` & just change the methods names.

I vote for `BitSize` or `BitsLength`

**Comment by enitrat** [comment]
I like both, but it's up to Ori to decide

Could you add an impl for `bytes31` @akhercha ? I just encountered a use case where it would be handy :)

**Comment by akhercha** [comment]
Since Orizi commented on my issue
> this is not "size" - this might be "representation bits" - "size" implies the size in memory.

I would go for `BitsLength` over `BitsSize`; that would give me:

```rust
trait BitsLength<T> {
    /// Returns the BitsLength in bits of T.
    fn bits() -> T;
    /// Returns the BitsLength in bits of self.
    fn bits_len(self: @T) -> T;
}
```

I can also rename it to `trait Length<T>` but not sure of this one... Any thoughts?

**Review by orizi — CHANGES_REQUESTED** [review]
Let us move this trait into:
`src/num/traits.cairo`

we'd have `One` and `Zero` there as well.
This qualification would make this much more readable.

Calling it simple `Bits` under this premise sounds good to me.

also `bits()` should probably return an index type (`usize`?) instead of `T`.

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251#-:-NgwZ00HEvNzyaKNERhe:b8trb0o)* status: 0 of 5 files reviewed, 3 unresolved discussions (waiting on @akhercha)

___
*[`corelib/src/bits_length.cairo` line 115 at r3](https://reviewable.io/reviews/starkware-libs/cairo/4251#-NgwYRMO5FSFyVQGwMUb:-NgwYRMO5FSFyVQGwMUc:b-6grnsk) ([raw file](https://github.com/starkware-libs/cairo/blob/ae9375494d771076436ac97b2e4b880011c90a51/corelib/src/bits_length.cairo#L115)):*
> ```Cairo
>     fn bits_len(self: @felt252) -> felt252 {
>         Felt252BitsLength::bits()
>     }
> ```

This is wrong - should just not be provided for felts.



_Code quote:_
```Cairo
impl Felt252BitsLength of BitsLength<felt252> {
    fn bits() -> felt252 {
        252
    }
    fn bits_len(self: @felt252) -> felt252 {
        Felt252BitsLength::bits()
    }
```

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 2 of 7 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: 3 of 8 files reviewed, 5 unresolved discussions (waiting on @akhercha)

___
*[`corelib/src/num/traits.cairo` line 5 at r5](https://reviewable.io/reviews/starkware-libs/cairo/4251#-NgxZEXR480Vpyvqfst4:-NgxZEXR480Vpyvqfst5:bcn2omi) ([raw file](https://github.com/starkware-libs/cairo/blob/92c0262d3e2c207e5e1c7a9117f8d4f4318cda3f/corelib/src/num/traits.cairo#L5)):*
> ```Cairo
> trait Bits<T> {
>     /// Returns the length in bits of T.
>     fn len() -> usize;
> ```

this is the correct name - and this is the main required function.

if we don't have a good name for the 



_Suggestion:_
```Cairo
    /// Returns the length in bits of T.
    fn bits() -> usize;
```
___
*[`corelib/src/num/traits.cairo` line 7 at r5](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Ngx_19i5l0kox6ea6Uo:-Ngx_19i5l0kox6ea6Up:b71ekm7) ([raw file](https://github.com/starkware-libs/cairo/blob/92c0262d3e2c207e5e1c7a9117f8d4f4318cda3f/corelib/src/num/traits.cairo#L7)):*
> ```Cairo
>     fn len() -> usize;
>     /// Returns the length in bits of self.
>     fn bits(self: @T) -> usize;
> ```

remove this for the time being - i don't see a particular need for this in the core lib.
adding an outside trait and impl extension providing this should be very easy - but rather not have it in the corelib for now.



_Code quote:_
```Cairo
    /// Returns the length in bits of self.
    fn bits(self: @T) -> usize;
```
___
*[`corelib/src/num/traits.cairo` line 111 at r5](https://reviewable.io/reviews/starkware-libs/cairo/4251#-NgxZ5w41XQHpTeMugOr:-NgxZ5w41XQHpTeMugOs:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/92c0262d3e2c207e5e1c7a9117f8d4f4318cda3f/corelib/src/num/traits.cairo#L111)):*
> ```Cairo
> impl Bytes31Bits of Bits<bytes31> {
>     fn len() -> usize {
>         31
> ```





_Suggestion:_
```Cairo
248
```

**Review by orizi — APPROVED** [review]
:lgtm:

> Reviewed 1 of 5 files at r3, 2 of 7 files at r4, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251#-:-NgyxWAz0TL85XxeuwN3:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @akhercha)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/4251#-NgyxSyp-ckomuT3BRBx:-NgyxSyp-ckomuT3BRBy:bpx0xxa) (no related file):*
@yuvalsw for 2nd eye.

___

**Review by yuvalsw — CHANGES_REQUESTED** [review]
> Reviewed 1 of 5 files at r3, 4 of 7 files at r4, 1 of 1 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: all files reviewed, 8 unresolved discussions (waiting on @akhercha)

___
*[`corelib/src/lib.cairo` line 338 at r6](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh26w5m4mwMw_VgkyiK:-Nh26w5n3BAlTo5PexG_:b-a7b47c) ([raw file](https://github.com/starkware-libs/cairo/blob/d040d6bca0ab7c188d09014fd279e450e8a14803/corelib/src/lib.cairo#L338)):*
> ```Cairo
> 
> // Num
> mod num;
> ```

Can you please rebase over https://reviewable.io/reviews/starkware-libs/cairo/4254 so we see the incremental diff?



_Code quote:_
```Cairo
// Num
mod num;
```
___
*[`corelib/src/num/traits.cairo` line 3 at r6](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh27FIZ5vQgh3HbR6Bv:-Nh27FIZ5vQgh3HbR6Bw:b8lfss) ([raw file](https://github.com/starkware-libs/cairo/blob/d040d6bca0ab7c188d09014fd279e450e8a14803/corelib/src/num/traits.cairo#L3)):*
> ```Cairo
> // === Bits ===
> 
> trait Bits<T> {
> ```

I actually liked `BitSize`/`BitLength` better. I think it's more descriptive and thus less confusing.
Between the 2, I prefer "size" as:
1. It's shorter.
2. Most documentation I know talk about size of types, not length (e.g. `size_of` in c/c++, `Sized` in Rust etc.).
3. It returns "usize".
4. If later a function with a self parameter for getting the size of its type would be added externally/internally: IMO "length" may sound like the actual length of the data in the current self, rather than the length of the type. "Size" somehow sounds to me more like the size of the type.

Anyway, please also doc the trait, especially following the relatively long naming discussion.
___
*[`corelib/src/num/traits.cairo` line 4 at r6](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh27EHY6m-XI1GcgZiZ:-Nh27EHY6m-XI1GcgZi_:b-7ilvfu) ([raw file](https://github.com/starkware-libs/cairo/blob/d040d6bca0ab7c188d09014fd279e450e8a14803/corelib/src/num/traits.cairo#L4)):*
> ```Cairo
> 
> trait Bits<T> {
>     /// Returns the bitsgth in bits of T.
> ```

typo?



_Code quote:_
```Cairo
bitsgth
```
___
*[`corelib/src/num/traits.cairo` line 8 at r6](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh28IPI9GuUgqekd4qF:-Nh28IPI9GuUgqekd4qG:bfb9iu9) ([raw file](https://github.com/starkware-libs/cairo/blob/d040d6bca0ab7c188d09014fd279e450e8a14803/corelib/src/num/traits.cairo#L8)):*
> ```Cairo
> }
> 
> impl U8Bits of Bits<u8> {
> ```

please move the impls to integer.cairo.
___
*[`corelib/src/test/num_test.cairo` line 10 at r6](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh29poyBQ-hkZzLawPn:-Nh29poyBQ-hkZzLawPo:b8tl8ax) ([raw file](https://github.com/starkware-libs/cairo/blob/d040d6bca0ab7c188d09014fd279e450e8a14803/corelib/src/test/num_test.cairo#L10)):*
> ```Cairo
> #[test]
> fn test_traits_bits() {
>     assert_eq(@(U8Bits::bits()), @8, 'U8 bits != 8');
> ```

aren't these parentheses redundant?



_Suggestion:_
```Cairo
@U8Bits::bits()
```

**Comment by akhercha** [comment]
- Updated the trait name to `BitSize`,
- Moved implementations to `integer.cairo` & `bytes_31.cairo`

I also rebased to `enitrat:feat/zero-one-traits` but I probably messed it up, my commit history is destroyed. 😅

**Review by orizi — COMMENTED** [review]
> Reviewed 1 of 7 files at r9.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: 5 of 15 files reviewed, 8 unresolved discussions (waiting on @akhercha and @yuvalsw)

**Review by yuvalsw — CHANGES_REQUESTED** [review]
> Reviewed 3 of 7 files at r8, 6 of 7 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: 13 of 15 files reviewed, 7 unresolved discussions (waiting on @akhercha and @orizi)

___
*[`corelib/src/bytes_31.cairo` line 4 at r9](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh5xOaTEdeRZqD3LKEs:-Nh5xOaTEdeRZqD3LKEt:b-tprmrm) ([raw file](https://github.com/starkware-libs/cairo/blob/ede86de6d2ccfcf092fea8026168fc3b5ced97fd/corelib/src/bytes_31.cairo#L4)):*
> ```Cairo
> use option::OptionTrait;
> use integer::{u256_from_felt252, u128_safe_divmod, u128_to_felt252};
> use num::traits::BitSize;
> ```

this also re-exports it from here. Better fully qualify in the usage instead.
___
*[`corelib/src/integer.cairo` line 7 at r9](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh5x_nN-Vc-DpuNbiDw:-Nh5x_nN-Vc-DpuNbiDx:b-vkv5n6) ([raw file](https://github.com/starkware-libs/cairo/blob/ede86de6d2ccfcf092fea8026168fc3b5ced97fd/corelib/src/integer.cairo#L7)):*
> ```Cairo
> use array::ArrayTrait;
> use array::SpanTrait;
> use num::traits::{Zero, One, BitSize};
> ```

same here, fully qualify instead.
___
*[`corelib/src/integer.cairo` line 2390 at r9](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh5zX1HCmhktt96PmYW:-Nh5zX1HCmhktt96PmYX:bxxtobc) ([raw file](https://github.com/starkware-libs/cairo/blob/ede86de6d2ccfcf092fea8026168fc3b5ced97fd/corelib/src/integer.cairo#L2390)):*
> ```Cairo
> }
> 
> // Zero trait implementations
> ```

please change target branch as well, so we only see the diff of this PR.

**Comment by akhercha** [comment]
*[`corelib/src/integer.cairo` line 2390 at r9](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh5zX1HCmhktt96PmYW:-Nh632gT7OITPjIGXb_C:b-i137nn) ([raw file](https://github.com/starkware-libs/cairo/blob/ede86de6d2ccfcf092fea8026168fc3b5ced97fd/corelib/src/integer.cairo#L2390)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

please change target branch as well, so we only see the diff of this PR.
</blockquote></details>

Do you have any guidance for how to do that properly? At this point I'm scared of breaking my branch again.. 😅

**Review by yuvalsw — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: 11 of 15 files reviewed, 7 unresolved discussions (waiting on @akhercha and @orizi)

___
*[`corelib/src/integer.cairo` line 2390 at r9](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh5zX1HCmhktt96PmYW:-Nh67gm-1QtzVlxBMqKF:bxysduc) ([raw file](https://github.com/starkware-libs/cairo/blob/ede86de6d2ccfcf092fea8026168fc3b5ced97fd/corelib/src/integer.cairo#L2390)):*
<details><summary><i>Previously, akhercha (akhercha) wrote…</i></summary><blockquote>

Do you have any guidance for how to do that properly? At this point I'm scared of breaking my branch again.. 😅
</blockquote></details>

AFAIK changing the target branch doesn't change anything about your branch. It just means that 1. The diff shown is based from the target branch. 2. If you merge, it's merged into this target (which you don't want - you want to have the target branch back to "main" after the other branch is merged and before merging yours).

Basically, you can hit "edit" in the github PR entry and change from main, or you can do it from revieweable by changing the "to" branch in the summary. But I am not able to do it - probably as I don't have access to the branch.
Another option is to wait for Eni's branch to be merged.

**Review by yuvalsw — COMMENTED** [review]
Thanks. Now that `enitrat:feat/zero-one-traits` was merged, please rebase back to main.

> Reviewed 2 of 2 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251#-:-NhgNvxC8jtIseXTJ1fO:bzte66)* status: 13 of 15 files reviewed, 5 unresolved discussions (waiting on @akhercha and @orizi)

___
*[`corelib/src/integer.cairo` line 2390 at r9](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh5zX1HCmhktt96PmYW:-NhgOH-g3m48S9xD_Uqm:ba6t30n) ([raw file](https://github.com/starkware-libs/cairo/blob/ede86de6d2ccfcf092fea8026168fc3b5ced97fd/corelib/src/integer.cairo#L2390)):*
<details><summary><i>Previously, akhercha (akhercha) wrote…</i></summary><blockquote>

Alright, thanks!
Looks like from both UIs I can't set the target to Eni's branch, it's not in the list. I'll just wait for his PR to be merged then!
</blockquote></details>

Yeah, I also couldn't do it, not sure why or how you should. Anyway, now it's merged. please re-rebase on main.

**Comment by akhercha** [comment]
*[`corelib/src/integer.cairo` line 2390 at r9](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nh5zX1HCmhktt96PmYW:-NhkFc9F3xsWAUIF36Ip:b-5kdvav) ([raw file](https://github.com/starkware-libs/cairo/blob/ede86de6d2ccfcf092fea8026168fc3b5ced97fd/corelib/src/integer.cairo#L2390)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

Yeah, I also couldn't do it, not sure why or how you should. Anyway, now it's merged. please re-rebase on main.
</blockquote></details>

Done! Should be good, let me know if anything needs to be changed.

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 3 of 12 files at r12, 5 of 9 files at r13, 4 of 4 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: 11 of 15 files reviewed, 3 unresolved discussions (waiting on @akhercha and @yuvalsw)

___
*[`corelib/src/num/traits/bit_size.cairo` line 5 at r14](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nhu_81O5BZuOs-rGqKf:-Nhu_81O5BZuOs-rGqKg:b-3a2if9) ([raw file](https://github.com/starkware-libs/cairo/blob/c449174eb03100fbb9487dfa345a007753ca4cdf/corelib/src/num/traits/bit_size.cairo#L5)):*
> ```Cairo
>     /// Returns the size in bits of T as usize.
>     fn bits() -> usize;
> }
> ```

i still don't like the discrepancy between function name and trait here.
any good ideas?  @yuvalsw 
I thought about something like `Bits::capacity`, and then later `Bits::width` (which is the actual number of used bits) could be added.
But i'm really not sure



_Code quote:_
```Cairo
/// Trait used to retrieve the size in bits of a type.
trait BitSize<T> {
    /// Returns the size in bits of T as usize.
    fn bits() -> usize;
}
```

**Review by yuvalsw — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: 11 of 15 files reviewed, 3 unresolved discussions (waiting on @akhercha and @orizi)

___
*[`corelib/src/num/traits/bit_size.cairo` line 5 at r14](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nhu_81O5BZuOs-rGqKf:-Ni-RvQXB3NEhCFgZozK:bl3fq0l) ([raw file](https://github.com/starkware-libs/cairo/blob/c449174eb03100fbb9487dfa345a007753ca4cdf/corelib/src/num/traits/bit_size.cairo#L5)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

i still don't like the discrepancy between function name and trait here.
any good ideas?  @yuvalsw 
I thought about something like `Bits::capacity`, and then later `Bits::width` (which is the actual number of used bits) could be added.
But i'm really not sure
</blockquote></details>

*IF* we want to add a method later for the actual number of used bits, `capacity` and `size` could be used for the function names. These are the terms generally used for such things (e.g. vector in c++).

But - I don't think we would ever want to add such a method. As I understand it, this is mainly aimed for basic types like u8, u16 etc, and I don't see the general use for a method for the actual number of bits used in such a basic type.
With that in mind - unless you think we would want to add such - I think the name `size` makes sense. If you want the trait name and the method name to be more consistent, I am also OK with changing the method name to `size`/`bit_size`. As there is not self parameter here, you must call it explicitly by BitSize::method_name(), so all are valid names IMO.

**Review by yuvalsw — APPROVED** [review]
> Reviewed 8 of 9 files at r13, 4 of 4 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: all files reviewed, 2 unresolved discussions (waiting on @akhercha and @orizi)

**Review by orizi — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251#-:-NiFSG-r9Eu4opBdGtR9:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @akhercha)

___
*[`corelib/src/num/traits/bit_size.cairo` line 5 at r14](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nhu_81O5BZuOs-rGqKf:-NiFRVEy1jwxjCT4vqt4:b-cwo3dk) ([raw file](https://github.com/starkware-libs/cairo/blob/c449174eb03100fbb9487dfa345a007753ca4cdf/corelib/src/num/traits/bit_size.cairo#L5)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

*IF* we want to add a method later for the actual number of used bits, `capacity` and `size` could be used for the function names. These are the terms generally used for such things (e.g. vector in c++).

But - I don't think we would ever want to add such a method. As I understand it, this is mainly aimed for basic types like u8, u16 etc, and I don't see the general use for a method for the actual number of bits used in such a basic type.
With that in mind - unless you think we would want to add such - I think the name `size` makes sense. If you want the trait name and the method name to be more consistent, I am also OK with changing the method name to `size`/`bit_size`. As there is not self parameter here, you must call it explicitly by BitSize::method_name(), so all are valid names IMO.
</blockquote></details>

the name `size` by itself is bad - as it is missing all context.
In anycase - resolving - let it be BitSize::bits();

**Review by orizi — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @akhercha)

**Review by yuvalsw — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251#-:-NifK3A04untP3JpQXBo:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @akhercha)

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 2 of 9 files at r13.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: all files reviewed, 1 unresolved discussion (waiting on @akhercha)

___
*[`corelib/src/test/num_test.cairo` line 24 at r14](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nj6afuB4AigcK-sbV2h:-Nj6afuC8jgYlpkV9mVr:b-272ime) ([raw file](https://github.com/starkware-libs/cairo/blob/c449174eb03100fbb9487dfa345a007753ca4cdf/corelib/src/test/num_test.cairo#L24)):*
> ```Cairo
>     assert_eq(@I128BitSize::bits(), @128, 'I128 bit size != 128');
>     assert_eq(@Bytes31BitSize::bits(), @248, 'Bytes31 bit size != 248');
> }
> ```

rebase and:



_Suggestion:_
```Cairo
use core::num::BitSize;

#[test]
fn test_bit_size() {
    assert!(BitSize::<u8>::bits() == 8);
    assert!(BitSize::<u16>::bits() == 16);
    assert!(BitSize::<u32>::bits() == 32);
    assert!(BitSize::<u64>::bits() == 64);
    assert!(BitSize::<u128>::bits() == 128);
    assert!(BitSize::<u256>::bits() == 256);
    assert!(BitSize::<i8>::bits() == 8);
    assert!(BitSize::<i16>::bits() == 16);
    assert!(BitSize::<i32>::bits() == 32);
    assert!(BitSize::<i64>::bits() == 64);
    assert!(BitSize::<i128>::bits() == 128);
    assert!(BitSize::<bytes31>::bits() == 248);
}
```

**Comment by akhercha** [comment]
*[`corelib/src/test/num_test.cairo` line 24 at r14](https://reviewable.io/reviews/starkware-libs/cairo/4251#-Nj6afuB4AigcK-sbV2h:-NjDuGP07TG0_VpM8QbQ:b-bw8e1h) ([raw file](https://github.com/starkware-libs/cairo/blob/c449174eb03100fbb9487dfa345a007753ca4cdf/corelib/src/test/num_test.cairo#L24)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

rebase and:
</blockquote></details>

Updated!

**Review by orizi — APPROVED** [review]
> Reviewed 8 of 8 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4251)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @akhercha)
